src/program/latex/mod.rs

```diff
@@ -0,0 +1,1600 @@
+        write!(f, "{}{}{}", self.header, self.body, self.footer)
+    }
+}
+
+/// A Diagram represents a collection of wires in a circuit. It encodes the
+/// wires row in the diagram and its relationship to other wires. A row is one
+/// of the wires in the circuit BTreeMap. Diagram tracks relationships between
+/// wires with two pieces of information--1. the wires row (its order in the
+/// BTreeMap), and 2. the column that spreads between all wires that pass
+/// through a multi qubit gate 'e.g. CNOT'. The size of the diagram can be
+/// measured by multiplying the column with the length of the circuit. This is
+/// an [m x n] matrix where each element in the matrix represents an item to be
+/// rendered onto the diagram using one of the [`Quantikz`] commands.
+#[derive(Debug)]
+struct Diagram {
+    /// customizes how the diagram renders the circuit
+    settings: Settings,
+    /// total number of elements on each wire
+    column: u32,
+    /// column at which qubits in positional order form relationships
+    relationships: HashMap<u32, Vec<u64>>,
+    /// a BTreeMap of wires with the name of the wire as the key
+    circuit: BTreeMap<u64, Box<Wire>>,
+}
+
+impl Default for Diagram {
+    fn default() -> Self {
+        Self {
+            settings: Settings::default(),
+            column: 0,
+            relationships: HashMap::new(),
+            circuit: BTreeMap::new(),
+        }
+    }
+}
+
+impl Diagram {
+    /// Compares qubits from a single instruction associated with a column on
+    /// the circuit to all of the qubits used in the quil program. If a qubit
+    /// from the quil program is not found in the qubits in the single
+    /// instruction line, then an empty slot is added to that column on the
+    /// qubit wire of the circuit.
+    ///
+    /// # Arguments
+    /// `&mut self` - exposes the Circuit
+    /// `qubits` - qubits used in the quil program
+    /// `instruction` - exposes qubits in a single instruction
+    fn set_qw(&mut self, qubits: &HashSet<Qubit>, instruction: &Instruction) {
+        for program_qubit in qubits {
+            let mut found = false;
+            match &instruction {
+                instruction::Instruction::Gate(gate) => {
+                    for gate_qubit in &gate.qubits {
+                        if program_qubit == gate_qubit {
+                            found = true;
+                        }
+                    }
+
+                    if !found {
+                        match program_qubit {
+                            instruction::Qubit::Fixed(q) => {
+                                self.circuit
+                                    .get_mut(&q)
+                                    .and_then(|wire| wire.empty.insert(self.column, Command::Qw));
+                            }
+                            _ => (),
+                        }
+                    }
+                }
+                _ => (),
+            }
+        }
```

Comment:
A few things here:

- If matching a single variant, prefer `if let` instead.
- Instead of using a boolean flag, use loop labels to `continue` the outer loop
- Style preference, `and_then` is more for `Option` combinatorics where you care about the returned value. In the case of this code, I think `if let Some(_)` is more appropriate

```suggestion
        'program_loop: for program_qubit in qubits {
            if let Instruction::Gate(gate) = instruction {
                for gate_qubit in &gate.qubits {
                    if program_qubit == gate_qubit {
                        continue 'program_loop;
                    }
                }

                if let Qubit::Fixed(q) = program_qubit {
                    if let Some(wire) = self.circuit.get_mut(&q) {
                        wire.empty.insert(self.column, Command::Qw);
                    }
                }
            }
        }
```